### PR TITLE
Migrate to symfony/console

### PR DESCRIPTION
**Pull request by weierophinney**
This patch migrates the tooling to use symfony/console instead of zend-stdlib's `ConsoleHelper`. The reasons are:

- We were needing to write argument parsing routines for each and every command, and each differed.
- We were needing to write "help" functionality for each command.
- We were creating a plethora of entry points. We should have one, with many commands available.
- While we could use zf-deploy, using symfony/console will likely get more contributors to the project, due to its ubiquity in the PHP ecosystem. In the end, I also discovered that I ended up writing less code overall in order to create commands, and had better, simpler options for re-use when re-purposing existing command endpoints to re-use the same commands as used by the main entry point.

All existing commands continue to work. I have updated the `expressive` tool to use the following patterns:

- `migrate:*`: the various tools dedicated to migration.
- `module:*`: the various tools dedicated to module manipulation.
- `middleware:*`: the tools dedicated to middleware manipulation
  (currently only creation).

### TODO

- [x] Update unit tests.
- [x] Add deprecation warnings to original, now legacy, scripts.

### CODE REVIEW

File: src/Module/CreateCommand.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        $modulesPath = $this->getModulesPath($input);
+
+        try {
+            $creation = new Create();
+            $message = $creation->process($module, $modulesPath, $this->projectDir);
+            $output->writeln(sprintf('<info>%s</info>', $message));
+        } catch (Exception\RuntimeException $e) {
+            $console = $this->getErrorConsole($output);
+            $console->writeln('<error>Error during execution:</error>', true, STDERR);
+            $console->writeln(sprintf('  <error>%s</error>', $ex->getMessage()), true, STDERR);
```
Comment (michalbundyra) [inline src/Module/CreateCommand.php]: `$ex` is not defined. Here should be `$e` or variable in catch should be renamed (I'd prefer to use `$ex` for exception)

File: src/Module/CreateCommand.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        $composer = $input->getOption('composer') ?: 'composer';
+        $modulesPath = $this->getModulesPath($input);
+
+        try {
+            $creation = new Create();
+            $message = $creation->process($module, $modulesPath, $this->projectDir);
+            $output->writeln(sprintf('<info>%s</info>', $message));
+        } catch (Exception\RuntimeException $e) {
+            $console = $this->getErrorConsole($output);
+            $console->writeln('<error>Error during execution:</error>', true, STDERR);
```
Comment (michalbundyra) [inline src/Module/CreateCommand.php]: Now we have different signature of `writeln` method - from symfony, so there are two params: `$messages` and `$options`.

File: src/Module/DeregisterCommand.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        $configProvider = sprintf('%s\ConfigProvider', $module);
+        if ($injector->isRegistered($configProvider)) {
+            $injector->remove($configProvider);
+        }
+
+        try {
+            $disable = new Disable($this->projectDir, $modulesPath, $composer);
+            $disable->process($module);
+        } catch (RuntimeException $ex) {
+            $console = $this->getErrorConsole();
```
Comment (michalbundyra) [inline src/Module/DeregisterCommand.php]: This method should be called with parameter. It's missing here.

File: src/Module/RegisterCommand.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                InjectorInterface::TYPE_CONFIG_PROVIDER
+            );
+        }
+
+        try {
+            $enable = new Enable($this->projectDir, $modulesPath, $composer);
+            $enable->setMoveModuleClass(false);
+            $enable->process($module);
+        } catch (RuntimeException $ex) {
+            $console = $this->getErrorConsole();
```
Comment (michalbundyra) [inline src/Module/RegisterCommand.php]: As above - missing required param.

File: src/ScanForErrorMiddleware/Scanner.php [hunk 5]
```diff
@@ -34,9 +34,9 @@ class Scanner implements Countable, IteratorAggregate
 
     /**
      * @param string $path
-     * @param ConsoleHelper $console
+     * @param OutputInterface $console
```
Comment (michalbundyra) [inline src/ScanForErrorMiddleware/Scanner.php:37]: The same should be changed also in line 21.

**Review by michalbundyra — CHANGES_REQUESTED** [review]
I like update to `symfony/console`! Many things is already done there so it simplify our library :) Couple of things to change and should be fine. Of course updating tests as you pointed is needed. Thanks!

**Comment by weierophinney** [comment]
All tests updated and/or new tests written; doing so addressed most of the feedback from @webimpress, and I've updated a property typehint per his feedback as well.

File: src/ScanForErrorMiddleware/Scanner.php (hunk 5 above)
Comment (weierophinney) [reply src/ScanForErrorMiddleware/Scanner.php:37]: Fixed.

**Comment by weierophinney** [comment]
Pinging @GeeH, @akrabat, and @CalEvans for feedback...

File: bin/expressive [hunk 6]
```diff
@@ -10,7 +10,9 @@
 
 namespace Zend\Expressive\Tooling;
 
-use Zend\Stdlib\ConsoleHelper;
+use Symfony\Component\Console\Application;
+
+const VERSION = '%version%';
```
Comment (Ocramius) [inline bin/expressive:15]: *cough* PackageVersions *cough*

File: bin/expressive [hunk 7]
```diff
@@ -22,6 +24,12 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
     exit(1);
 }
 
-$command = new ExpressiveCommand($argv[0], new ConsoleHelper());
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive', VERSION);
+$application->add(new CreateMiddleware\CreateMiddlewareCommand('middleware:create'));
```
Comment (Ocramius) [inline bin/expressive]: [`Application#addCommands()`](https://github.com/symfony/symfony/blob/bbb0d5ed424a2215588b65720434a0de34611ea0/src/Symfony/Component/Console/Application.php#L399)

File: bin/expressive-migrate-original-messages [hunk 8]
```diff
@@ -22,6 +24,6 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
     exit(1);
 }
 
-$command = new Command($argv[0], new ConsoleHelper());
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive-migrate-original-messages', VERSION);
```
Comment (Ocramius) [inline bin/expressive-migrate-original-messages]: Can include the main CLI app, then set the default command instead.

File: bin/expressive-module [hunk 9]
```diff
@@ -22,6 +24,8 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
     exit(1);
 }
 
-$command = new Command($argv[0], new ConsoleHelper());
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive-module', VERSION);
+$application->add(new CreateCommand('create'));
```
Comment (Ocramius) [inline bin/expressive-module]: Was this done just to keep the command names BC compliance?

File: bin/expressive-pipeline-from-config [hunk 10]
```diff
@@ -22,6 +24,6 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
     exit(1);
 }
 
-$command = new Command($argv[0], new ConsoleHelper());
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive-pipeline-from-config', VERSION);
```
Comment (Ocramius) [inline bin/expressive-pipeline-from-config]: Same as above: main app, set default command in this file

File: bin/expressive-scan-for-error-middleware [hunk 11]
```diff
@@ -22,6 +24,6 @@ if (file_exists($a = __DIR__ . '/../../../autoload.php')) {
     exit(1);
 }
 
-$command = new Command($argv[0], new ConsoleHelper());
-$return = $command->process(array_slice($argv, 1));
-exit($return);
+$application = new Application('expressive-scan-for-error-middleware', VERSION);
```
Comment (Ocramius) [inline bin/expressive-scan-for-error-middleware]: Same as above: main app, set default command in this file

File: composer.json [hunk 12]
```diff
@@ -43,7 +47,6 @@
     },
     "bin": [
         "bin/expressive",
-        "bin/expressive-create-middleware",
```
Comment (Ocramius) [inline composer.json:46]: Possible BC break? Or do we not want to go there?

File: composer.json [hunk 13]
```diff
@@ -9,13 +9,17 @@
...
+        "sort-packages": true
+    },
     "extra": {
         "branch-alias": {
             "dev-master": "1.0-dev"
         }
     },
     "require": {
         "php": "^5.6 || ^7.0",
+        "symfony/console": "^3.2",
```
Comment (Ocramius) [inline composer.json]: Please also specify the current patch version used: we don't want people to start with 3.2.0, but rather 3.2.7 (current)

File: src/ErrorConsoleTrait.php [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+trait ErrorConsoleTrait
+{
+    /**
+     * Retrieve console output, if possible, to allow writing to STDERR.
+     *
+     * @param OutputInterface $output
+     * @return OutputInterface
+     */
+    private function getErrorConsole(OutputInterface $output)
```
Comment (Ocramius) [inline src/ErrorConsoleTrait.php]: All this fuss just for some errors? Nah, just let the exceptions bubble up, and consumers can use `-vvv` to see all the details

File: src/GenerateProgrammaticPipelineFromConfig/Generator.php [hunk 15]
```diff
@@ -240,10 +239,11 @@ private function generatePipeline(array $config)
             $error      = isset($spec['error']) ? (bool) $spec['error'] : false;
 
             if ($error) {
-                $this->console->writeLine(sprintf(
+                $console = $this->getErrorConsole($this->console);
```
Comment (Ocramius) [inline src/GenerateProgrammaticPipelineFromConfig/Generator.php]: Would suggest just printing to `$this->console`, unless STDOUT is important in this generator

File: src/ErrorConsoleTrait.php (hunk 14 above)
Comment (Ocramius) [reply src/ErrorConsoleTrait.php]: If really needed, can be made into a `public static function` instead (better, and testable)

File: src/CreateMiddleware/CreateMiddlewareCommand.php [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+        $generator = new CreateMiddleware();
+
+        try {
+            $path = $generator->process($middleware, $this->projectDir);
+        } catch (CreateMiddlewareException $e) {
+            $console = $this->getErrorConsole($output);
+            $console->writeln('<error>Error during generation:</error>');
+            $console->writeln(sprintf('  <error>%s</error>', $e->getMessage()));
+            return 1;
```
Comment (Ocramius) [inline src/CreateMiddleware/CreateMiddlewareCommand.php]: Let the exception bubble up, otherwise you are making it really hard for consumers to see why the `$generator` may have failed, as no stack trace is produced

File: src/GenerateProgrammaticPipelineFromConfig/PipelineFromConfigCommand.php [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+
+        try {
+            $generator = new Generator($output);
+            $generator->projectDir = $this->projectDir;
+            $generator->process($this->locateConfigFile($input));
+        } catch (GeneratorException $e) {
+            $console = $this->getErrorConsole($output);
+            $console->writeln('<error>Error during generation:</error>');
+            $console->writeln(sprintf('  <error>%s</error>', $e->getMessage()));
+            return 1;
```
Comment (Ocramius) [inline src/GenerateProgrammaticPipelineFromConfig/PipelineFromConfigCommand.php]: Same: let the exception bubble up instead

File: src/MigrateOriginalMessageCalls/MigrateOriginalMessageCallsCommand.php [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    {
+        try {
+            $src = $this->getSrcDir($input);
+        } catch (ArgvException $e) {
+            $console = $this->getErrorConsole($output);
+            $console->writeln(sprintf(
+                '<error>Unable to determine src directory: %s</error>',
+                $e->getMessage()
+            ));
+            return 1;
```
Comment (Ocramius) [inline src/MigrateOriginalMessageCalls/MigrateOriginalMessageCallsCommand.php]: Let exceptions bubble up. If the exception message is meaningful, no additional code is needed

File: src/Module/CreateCommand.php [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        $module = $input->getArgument('module');
+        $composer = $input->getOption('composer') ?: 'composer';
+        $modulesPath = $this->getModulesPath($input);
+
+        try {
+            $creation = new Create();
+            $message = $creation->process($module, $modulesPath, $this->projectDir);
+            $output->writeln(sprintf('<info>%s</info>', $message));
+        } catch (Exception\RuntimeException $e) {
```
Comment (Ocramius) [inline src/Module/CreateCommand.php]: remove, let it bubble up

File: src/Module/DeregisterCommand.php [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+        $injector = new ConfigAggregatorInjector($this->projectDir);
+        $configProvider = sprintf('%s\ConfigProvider', $module);
+        if ($injector->isRegistered($configProvider)) {
+            $injector->remove($configProvider);
+        }
+
+        try {
+            $disable = new Disable($this->projectDir, $modulesPath, $composer);
+            $disable->process($module);
+        } catch (RuntimeException $ex) {
```
Comment (Ocramius) [inline src/Module/DeregisterCommand.php]: Remove, let it bubble up

File: src/Module/RegisterCommand.php [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+                $configProvider,
+                InjectorInterface::TYPE_CONFIG_PROVIDER
+            );
+        }
+
+        try {
+            $enable = new Enable($this->projectDir, $modulesPath, $composer);
+            $enable->setMoveModuleClass(false);
+            $enable->process($module);
+        } catch (RuntimeException $ex) {
```
Comment (Ocramius) [inline src/Module/RegisterCommand.php]: same

File: src/ScanForErrorMiddleware/ScanForErrorMiddlewareCommand.php [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+     *
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return int Exit status
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        try {
+            $src = $this->getSrcDir($input);
+        } catch (ArgvException $e) {
```
Comment (Ocramius) [inline src/ScanForErrorMiddleware/ScanForErrorMiddlewareCommand.php]: same

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: bin/expressive (hunk 6 above)
Comment (weierophinney) [reply bin/expressive:15]: *cough* PHP 5.6 *cough*

File: bin/expressive (hunk 7 above)
Comment (weierophinney) [reply bin/expressive]: Cool, wasn't aware of that one!

File: bin/expressive-migrate-original-messages (hunk 8 above)
Comment (weierophinney) [reply bin/expressive-migrate-original-messages]: This is a BC measure; the goal is to eventually remove it.

Also note that the command names differ between the main `expressive` script and this one.

File: bin/expressive-module (hunk 9 above)
Comment (weierophinney) [reply bin/expressive-module]: Yes. Since we already have documented migration steps, I cannot break how these original scripts worked; hence the difference between command names and what commands are exposed.

File: bin/expressive-pipeline-from-config (hunk 10 above)
Comment (weierophinney) [reply bin/expressive-pipeline-from-config]: See above; this is a BC measure.

File: bin/expressive-scan-for-error-middleware (hunk 11 above)
Comment (weierophinney) [reply bin/expressive-scan-for-error-middleware]: See above; this is a BC measure.

File: composer.json (hunk 12 above)
Comment (weierophinney) [reply composer.json:46]: Never before released; that feature was for the upcoming 0.4.0 release. As such, I'd rather _not_ release that script, and instead ship it only as a command in the `expressive` script.

**Comment by GeeH** [comment]
The move to this library is amazing, we should always look to use community defacto libraries instead of having to code and support our own whatever the vendor namespace. I'm happy.

I don't have time to review the code sorry (and I don't know the library very well anyway), but...

image

**Comment by weierophinney** [comment]
@Ocramius I've incorporated your feedback; thanks a ton!

File: src/GenerateProgrammaticPipelineFromConfig/PipelineFromConfigCommand.php [hunk 23]
```diff
@@ -64,16 +61,9 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
-            $generator->projectDir = $this->projectDir;
-            $generator->process($this->locateConfigFile($input));
-        } catch (GeneratorException $e) {
-            $console = $this->getErrorConsole($output);
-            $console->writeln('<error>Error during generation:</error>');
-            $console->writeln(sprintf('  <error>%s</error>', $e->getMessage()));
-            return 1;
-        }
+        $generator = new Generator($output);
+        $generator->projectDir = $this->projectDir;
```
Comment (Ocramius) [inline src/GenerateProgrammaticPipelineFromConfig/PipelineFromConfigCommand.php]: Argh... Missed this before, why is this necessary?

File: src/Module/CommandCommonTrait.php [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+{
+    /**
+     * @var string Root of project.
+     */
+    public $projectDir = '.';
+
+    /**
+     * Add default arguments and options used by all commands.
+     */
+    private function addDefaultOptionsAndArguments()
```
Comment (Ocramius) [inline src/Module/CommandCommonTrait.php]: This should be a `public static function`, referenced by classes using it.

File: src/Module/CommandCommonTrait.php [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+        );
+    }
+
+    /**
+     * Retrieve the modules path from input
+     *
+     * @param InputInterface $input
+     * @return string
+     */
+    private function getModulesPath(InputInterface $input)
```
Comment (Ocramius) [inline src/Module/CommandCommonTrait.php]: This should be a `public static function`, referenced by classes using it.

File: src/Module/CommandCommonTrait.php [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+ * @license   https://github.com/zendframework/zend-expressive-tooling/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Expressive\Tooling\Module;
+
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+
+trait CommandCommonTrait
```
Comment (Ocramius) [inline src/Module/CommandCommonTrait.php]: To be moved to a static class and marked as `@internal`, since we really don't want to distribute this as public/reusable API

**Review by michalbundyra — APPROVED** [review]

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/GenerateProgrammaticPipelineFromConfig/PipelineFromConfigCommand.php (hunk 23 above)
Comment (weierophinney) [reply src/GenerateProgrammaticPipelineFromConfig/PipelineFromConfigCommand.php]: Actually... no longer necessary. It was necessary before I started mocking the various internal classes, but now it isn't. I'll nuke it in the various `*Command` classes.

It's still necessary for the internal classes that do filesystem operations, though I _may_ be able to move that to constructor arguments. I'll review.

File: src/Module/CommandCommonTrait.php (hunk 26 above)
Comment (weierophinney) [reply src/Module/CommandCommonTrait.php]: Fair enough; I'll make that change (and the ones you suggest below) momentarily.

**Review by Ocramius — APPROVED** [review]

**Comment by Ocramius** [comment]
:ship:
